### PR TITLE
[release/3.1] Use Sse2 instrinsics to optimize NeedsEscaping/FindFirstCharToEncode for all built-in JavaScriptEncoders

### PR DESCRIPTION
**Pull request by ahsonkhan**
Port the set of performance optimizations related to escaping checks: https://github.com/dotnet/corefx/pull/41845, https://github.com/dotnet/corefx/pull/41933, https://github.com/dotnet/corefx/pull/42023 and add necessary fixes that are dependent on the release/3.1 branch (such as use of TFM based constants instead of the SDK one)

The PR descriptions highlight the performance improvements. Here's a gist with the analysis:
https://gist.github.com/ahsonkhan/c566f5e7d65c1fde5a83a67be290c4ee

### Description

No functional/behavioral change is intended by this change, other than perf optimizations.

Here are the key improvements:
- Use Sse2 intrinsics to process 8 characters at a time, otherwise fall back to the sequential code path
- Use a int[] cache to store the ASCII characters as they get processed for better caching perf
- Loop unrolling in areas where perf matters and it was feasible

This set of optimizations was applied to all the built-in concrete encoders (`Default`, `UnsafeRelaxedJsonEscaping`, ones created from the `Create` factory method, and the `TextEncoder` virtual methods), and since it was applied on multiple types, some of this change involves refactoring out the common code into helpers/separate files.

### Customer Impact:

Callers of the S.T.Json `JsonSerializer` and `Utf8JsonWriter` get faster when using any encoder options, particularly for writing strings that are large (>= 16 characters). This change also targets improving users of ASP.NET since they set the `JsonSerializerOption` to use the non-default `JavaScriptEncoder` by default (`UnsafeRelaxedJsonEscaping`).

Here's a summary of the results (regardless of whether folks use the default encoder or pass in the custom built-in encoders via the `JsonSerializerOptions` or `JsonWriterOptions`).
1) For end-to-end scenario (such as serializing commonly found objects/payloads), there is a 10-15% improvement. This is particularly noticeable for an object model/payload used by the NuGet client team (`SearchResult`).
2) Writing relatively large JSON strings using the writer got ~30% faster (i.e. greater than 16 characters).
3) Checking to escape UTF-16 (for strings of length >= 16)
   - using Default improved 2x
   - using Relaxed improved 3x
   - using Custom improved 15-20%
   - small regression for lengths <= 4.
4) Checking to escape UTF-8 (for strings of length >= 16)
   - using Default improved 2-5x
   - using Relaxed improved 6-10x
   - using Custom improved 2-3x
   - small regression for lengths <= 2.

By moving the "NeedsEscaping" logic to its correct location - S.T.E.Web and removing duplicate code from S.T.Json leads to better separation of concerns/single responsibility and helps with maintenance by mitigating potential bugs that could occur when we have duplicate logic.

### Regression? 

No

### Risk

The primary risk with this change is the escaping check breaks for some edge case/length. This is mitigated by code review from multiple folks and exhaustive tests (test coverage of the added code is ~100%). The code was also written defensively using Debug.Asserts aggressively to reason about the invariant. Given the performance gains is substantial, and moving the logic into S.T.E.W follows better encapsulation, I think the risk is justified.

### Tests run / added

A big part of this change includes adding tests for various encoders and input strings to validate that the right index is being returned for which character needs to be escaped. This includes surrogate pairs, invalid strings, empty strings, etc.

cc @GrabYourPitchforks, @steveharter, @ericstj, @danmosemsft

### CODE REVIEW

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoder.cs [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+using System.Text.Unicode;
+
+#if BUILDING_INBOX_LIBRARY
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+#endif
+
+namespace System.Text.Encodings.Web
+{
+    internal sealed class DefaultJavaScriptEncoder : JavaScriptEncoder
```
Comment (ahsonkhan) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoder.cs]: Most of the changes in this file are copy-paste from `JavaScriptEncoder.cs` into a separate file and don't change the existing logic.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/JavaScriptEncoderHelper.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Text.Unicode;
+
+namespace System.Text.Encodings.Web
+{
+    internal static class JavaScriptEncoderHelper
```
Comment (ahsonkhan) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/JavaScriptEncoderHelper.cs]: This file is extracting out common code into helpers to avoid duplication but is unchanged from what's already there.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+                        goto Return;
+                    }
+                    idx += 8;
+                    startingAddress += 8;
+                }
+
+                // Process the remaining characters.
+                Debug.Assert(textLength - idx < 8);
+            }
+#endif
```
Comment (ahsonkhan) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs]: This is the crux of the optimization (applied to both utf-16 and utf-8 overloads for the different `JavaScriptEncoders`).

File: src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs [hunk 4]
```diff
@@ -52,60 +54,24 @@ internal static partial class JsonWriterHelper
...
 
         public static unsafe int NeedsEscaping(ReadOnlySpan<char> value, JavaScriptEncoder encoder)
         {
-            int idx;
-
-            // Some implementations of JavascriptEncoder.FindFirstCharacterToEncode may not accept
-            // null pointers and gaurd against that. Hence, check up-front and fall down to return -1.
-            if (encoder != null && !value.IsEmpty)
+            // Some implementations of JavaScriptEncoder.FindFirstCharacterToEncode may not accept
+            // null pointers and gaurd against that. Hence, check up-front to return -1.
```
Comment (Tornhoof) [inline src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs]: Tiny Typo: gaurd -> guard.

**Comment by danmosemsft** [comment]
Not for Preview 2.  Seems too large for 3.1 but let's talk next week. Great change for master.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoder.cs [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                        Debug.Assert(nextScalarValue <= int.MaxValue);
+                        if (opStatus != OperationStatus.Done || WillEncode((int)nextScalarValue))
+                        {
+                            goto Return;
+                        }
+
+                        Debug.Assert(opStatus == OperationStatus.Done);
+                        idx += utf8BytesConsumedForScalar;
+                    }
+                }
```
Comment (tarekgh) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoder.cs]: does it make sense to add assert after this line idx == utf8Text.Length?

Comment (ahsonkhan) [reply src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoder.cs]: The while loop condition already guards for `idx < utf8Text.Length`, so adding the assert would be primarily to make sure idx isn't greater than utf8Text.Length. Sure, we can add it.

**Comment by ahsonkhan** [comment]
Closing this. I'll re-evaluate if we want this change in 3.1.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoder.cs (hunk 5 above)
Comment (ahsonkhan) [reply src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoder.cs]: Addressed this nit in https://github.com/dotnet/corefx/pull/42064
